### PR TITLE
Use OkHttp library for concurrent url resolving

### PR DESCRIPTION
**Pull request by rtyley**
We saw a lot of `IOException("too many concurrent streams")` errors in the AWS Lambda from PR https://github.com/guardian/google-search-indexing-observatory/pull/12 where we were using the `java.net.http HTTPClient` (added in Java 11) for resolving lots of urls at once - we had to revert it. According to [one answer on stackoverflow](https://stackoverflow.com/a/57359328/438886), the OkHTTP library behaves better:

> "Finally, I ended up by using OkHttp. It handles the problem (it just waits until some streams are freed up if the number of concurrent streams reaches max_concurrent_streams)."

This PR follows on from https://github.com/guardian/google-search-indexing-observatory/pull/37, which committed most of the redirect-resolving code, but deliberately omitted _calling_ the redirecting code, as it was throwing those `IOException`s.

We've verified over the past 2 days (Wednesday 4th-Friday 6th October) with a [test deploy](https://riffraff.gutools.co.uk/deployment/view/4250b546-7034-40eb-83a7-c7e6e0694a1f) that there are [no exceptions being thrown](https://logs.gutools.co.uk/s/ophan/goto/3dd0cf00-6450-11ee-bdfd-cb86db0ad0a5) on the AWS Lambda when we use OkHttp rather than Java 11's `java.net.http HTTPClient` (as used in commit 6e1f88ef022150dd50414440e7ffa84cc5ac8b55):

image

### CODE REVIEW

File: src/main/scala/ophan/google/indexing/observatory/RedirectResolver.scala [hunk 1]
```diff
@@ -60,32 +55,40 @@ object RedirectFollower extends RedirectFollower with Logging {
...
+      .build();
 
-
-  def follow(uri: URI): Future[Resp] = httpClient.sendAsync(requestFor(uri), BodyHandlers.discarding()).asScala.map {
+  def follow(uri: URI): Future[Resp] = Future {
+    blocking {
+      httpClient.newCall(requestFor(uri)).execute()
+    }
+  }.map {
     response => redirectGivenBy(uri, response).toRight {
```
Comment (rtyley) [inline src/main/scala/ophan/google/indexing/observatory/RedirectResolver.scala:90]: OkHTTP does have some support for Async requests, not not using the Scala-compatible Java `CompletionFuture`s:

https://square.github.io/okhttp/recipes/#asynchronous-get-kt-java

The recipe was slightly more complicated, so we dodged putting that into a Scala `Future` for now, and instead blocked a thread for each request - this probably has it's own throttling effect of some kind.

**Comment by rtyley** [comment]
We did a test [deploy](https://riffraff.gutools.co.uk/deployment/view/4250b546-7034-40eb-83a7-c7e6e0694a1f) of this yesterday with build 113, and the logs show that it is (mostly) working well, with no exceptions being thrown:

image

### Logging problem

The one thing that's _not_ working is the logging that tells us how many new urls are being discovered/resolved/stored-to-dynamodb:

https://github.com/guardian/google-search-indexing-observatory/blob/da98b2d0e5b6c1ee41315facf0bcce97f96b7a0e/src/main/scala/ophan/google/indexing/observatory/DataStore.scala#L36-L42

It's going into Cloudwatch Logs...

```
{
    "@timestamp": "2023-10-04T10:34:26.459459Z",
    "@version": "1",
    "message": "Storing new uris for https://www.dailymail.co.uk",
    "logger_name": "ophan.google.indexing.observatory.DataStore",
    "thread_name": "scala-execution-context-global-21",
    "level": "INFO",
    "level_value": 20000,
    "sitemap.uris.all": 1000,
    "sitemap.uris.new.resolved.notOK.sample": [],
    "sitemap.uris.new.resolved.sample": [
        "https://www.dailymail.co.uk/femail/article-12592417/Cosy-Kate-Princess-Wales-dons-grey-knitted-vest-visit-community-hub-supporting-Ukrainian-refugees.html"
    ],
    "sitemap.uris.new.resolved.notOK.count": 0,
    "sitemap.uris.new.redirecting.count": 0,
    "sitemap.uris.new.redirecting.sample": [],
    "sitemap.uris.new.resolved.count": 1,
    "site": "https://www.dailymail.co.uk",
    "buildNumber": "113",
    "gitCommitId": "381488d81adc9469f85999e9585b807205241ffe",
    "uniqueIdForVM": "f4554b17-119c-4200-8c86-bde7461bb5f5"
}
```

...but not making it into the [ELK logs](https://logs.gutools.co.uk/s/ophan/goto/4e355590-62a8-11ee-87de-730dde97d490). This issue was very similar to https://github.com/guardian/ophan/pull/3254 & https://github.com/guardian/google-search-indexing-observatory/pull/35 (tho' this last one was easier to resolve, as the document we were constructing was inconsistent with itself, which wasn't the case here)

This [conversation on DevX](https://chat.google.com/room/AAAAag0I08g/IIFm9n7snTM) clarified the issue - Louis Hather was able to dig out this relevant log line from the logstash box itself (note that unfortunately these errors are not available in the ELK itself, so you probably have to SSM onto the logstash box to find them)

```
Could not dynamically add mapping for field [sitemap.uris.new.redirecting.sample]. Existing mapping for [sitemap.uris.new] must be of type object but found [long]
```

I found this error message hard to parse, but Andrew Nowak pointed out that it effectively means:

> "You're asking me to store an **object** in a field called `sitemap.uris.new`, but that field is defined as a `long` in the existing mapping for the Elasticsearch index, so that object can't go in"

We can see the field, being used a `long`, here in an old log entry:

image

That old log entry was running with commit 6edd2fbaf5421c75117de3fb2f12a52313ad79e3, when indeed an integer was being logged to that field:

https://github.com/guardian/google-search-indexing-observatory/blob/6edd2fbaf5421c75117de3fb2f12a52313ad79e3/src/main/scala/ophan/google/indexing/observatory/DataStore.scala#L35

I'll fix the logging and do another test deploy.

**Comment by rtyley** [comment]
The fixed logging has been [deployed](https://riffraff.gutools.co.uk/deployment/view/579807ca-ecfd-4722-bc1a-849af484a3d9) as build 115, and the [logging for storing new uris](https://logs.gutools.co.uk/s/ophan/goto/47962fa0-6381-11ee-9e7f-759f013f1f57) is working well:

image

The [logging is showing the redirecting code correctly working](https://logs.gutools.co.uk/s/ophan/goto/8b13b680-6381-11ee-aae5-758248b644b6), identifying redirecting urls like https://www.bbc.co.uk/news/world-us-canada-67011628 :

image

There has been one url that didn't resolve, https://www.bbc.co.uk/news/live/67018505 returned a HTTP 500 error when we tried to resolve it ([logs](https://logs.gutools.co.uk/s/ophan/goto/2ef12d00-6382-11ee-9e7f-759f013f1f57)) - this rate of failure seems ok to me so far:

image

The specific aim for this work (originally started in https://github.com/guardian/google-search-indexing-observatory/pull/12) was to fix scanning of urls which redirect https://github.com/guardian/google-search-indexing-observatory/issues/32, which I think is done!

More broadly, we want the observatory to accurately reflect reality, and _not_ report items as missing from Google Search when they're actually there - that makes it look like Google Search is worse than reality, and alarms people for no good reason. I'll start a new issue to reflect one particular thing I've seen.

File: src/main/scala/ophan/google/indexing/observatory/AvailabilityUpdaterService.scala [hunk 2]
```diff
@@ -35,8 +36,8 @@ case class AvailabilityUpdaterService(
   def processNewUris(sitemapDownload: SitemapDownload, excludingAlreadyExistingRecords: Set[AvailabilityRecord]): Future[_] = {
     val urisNotSeenBefore = sitemapDownload.allUris -- excludingAlreadyExistingRecords.map(_.uri)
     for {
-      // redirectResolutionsForUrisNotSeenBefore <- Future.traverse(urisNotSeenBefore)(redirectResolver.resolve)
-      _ <- dataStore.storeNewRecordsFor(sitemapDownload, urisNotSeenBefore)
+      redirectResolutionsForUrisNotSeenBefore <- Future.traverse(urisNotSeenBefore)(redirectResolver.resolve)
+      _ <- dataStore.storeNewRecordsFor(sitemapDownload, redirectResolutionsForUrisNotSeenBefore)
```
Comment (rtyley) [inline src/main/scala/ophan/google/indexing/observatory/AvailabilityUpdaterService.scala:40]: Here we're re-instating the call to the redirect-resolving code.

File: src/main/scala/ophan/google/indexing/observatory/RedirectResolver.scala [hunk 3]
```diff
@@ -60,32 +54,39 @@ object RedirectFollower extends RedirectFollower with Logging {
...
 
-  def requestFor(uri: URI): HttpRequest =
-    HttpRequest.newBuilder().uri(uri).timeout(ofSeconds(2))
-      .header("User-Agent", "curl/7.54") // NYT accepts curl, but rejects the default "Java-http-client" User-Agent!
-      .method("HEAD", noBody()).build()
+  def requestFor(uri: URI): Request =
+    new Request.Builder()
+      .url(uri.toURL)
+      .head()
+      .header("User-Agent", "curl/7.54")
```
Comment (DavidLawes) [inline src/main/scala/ophan/google/indexing/observatory/RedirectResolver.scala:80]: Would it be worth copying the comment across (about NYT rejecting the default java http client user agent) or is it no longer applicable?

File: src/main/scala/ophan/google/indexing/observatory/model/Site.scala [hunk 4]
```diff
@@ -22,7 +22,7 @@ object Sites {
   )
 
   val BBC = Site(
-    url = "https://www.bbc.co.uk",
+    url = "https://www.bbc.co.uk/",
```
Comment (DavidLawes) [inline src/main/scala/ophan/google/indexing/observatory/model/Site.scala:25]: Is the trailing `/` significant in any way?

**Review by DavidLawes — APPROVED** [review]
What an exciting project to learn about, thanks!

File: src/main/scala/ophan/google/indexing/observatory/RedirectResolver.scala (hunk 3 above)
Comment (rtyley) [reply src/main/scala/ophan/google/indexing/observatory/RedirectResolver.scala:80]: Good question! I'm kind of in two minds about it, but the reason I decided against the comment there is that there's actually also a test committed that carries the same information:

https://github.com/guardian/google-search-indexing-observatory/blob/7cf7dcb430d7330e8b08fd2a81ba1d3c33052ecf/src/test/scala/ophan/google/indexing/observatory/RedirectResolverTest.scala#L19-L25

A **test** is often better than a **comment**, because a comment can fall out of date, no longer be true (yet _seem_ authoritative, and therefore be actively misleading) - but a test will actually tell us if what we say in the comment is true - does adding 'curl' to the user-agent allow us to resolve a NYT article? And if some decides, hey, why is this `.header("User-Agent", "curl/7.54")` line of code there - let's remove it! - the tests will quickly tell them why that's not a good idea.

Doing a git blame on the line will probably also get them the information, which is also good, because then it specifically places it in context as a-thing-that-was-true-at-a-point-in-time, rather than the eternal this-is-true-right-now vibes that an inline code comment gives.

File: src/main/scala/ophan/google/indexing/observatory/model/Site.scala (hunk 4 above)
Comment (rtyley) [reply src/main/scala/ophan/google/indexing/observatory/model/Site.scala:25]: I just realised, when looking at the logging, that the site urls were inconsistent - some had ending slashes and others didn't! So I just made them consistent.

File: src/main/scala/ophan/google/indexing/observatory/RedirectResolver.scala (hunk 3 above)
Comment (DavidLawes) [reply src/main/scala/ophan/google/indexing/observatory/RedirectResolver.scala:80]: Ah, amazing - I'd missed the test, thanks :)
